libcudacxx/include/cuda/std/__random/philox_engine.h

```diff
@@ -0,0 +1,558 @@
+          ::cuda::std::size_t _WordSize,
+          ::cuda::std::size_t _WordCount,
+          ::cuda::std::size_t _NumRounds,
+          _UIntType... _Constants>
+class philox_engine
+{
+  static_assert(::cuda::std::is_unsigned_v<_UIntType>, "philox_engine: _UIntType must be an unsigned integer type");
+  static_assert(_WordCount == 2 || _WordCount == 4, "N argument must be either 2 or 4");
+  static_assert(sizeof...(_Constants) == _WordCount, "consts array must be of length N");
+  static_assert(_NumRounds > 0, "rounds must be a strictly positive number");
+  static_assert((0 < _WordSize && _WordSize <= ::cuda::std::numeric_limits<_UIntType>::digits),
+                "Word size w must satisfy 0 < w <= numeric_limits<_UIntType>::digits");
+  static _CCCL_API constexpr auto __multipliers()
+  {
+    constexpr _UIntType __constants[] = {_Constants...};
+    if constexpr (_WordCount == 2)
+    {
+      return ::cuda::std::array<_UIntType, 1>{__constants[0]};
+    }
+    else
+    {
+      return ::cuda::std::array<_UIntType, 2>{__constants[0], __constants[2]};
+    }
+  }
+  static _CCCL_API constexpr auto __round_consts()
+  {
+    constexpr _UIntType __constants[] = {_Constants...};
+    if constexpr (_WordCount == 2)
+    {
+      return ::cuda::std::array<_UIntType, 1>{__constants[1]};
+    }
+    else
+    {
+      return ::cuda::std::array<_UIntType, 2>{__constants[1], __constants[3]};
+    }
+  }
+
+public:
+  using result_type                                = _UIntType;
+  static constexpr ::cuda::std::size_t word_size   = _WordSize;
+  static constexpr ::cuda::std::size_t word_count  = _WordCount;
+  static constexpr ::cuda::std::size_t round_count = _NumRounds;
+  static constexpr auto multipliers                = __multipliers();
+  static constexpr auto round_consts               = __round_consts();
+  static constexpr result_type default_seed        = 20111115u;
+
+  //! The smallest value this engine may potentially produce.
+  [[nodiscard]] _CCCL_API static constexpr result_type min() noexcept
+  {
+    return 0;
+  }
+  //! The largest value this engine may potentially produce.
+  [[nodiscard]] _CCCL_API static constexpr result_type max() noexcept
+  {
+    return ((1ull << (word_size - 1)) | ((1ull << (word_size - 1)) - 1));
+  }
+
+  //! This constructor, which optionally accepts a seed, initializes a new
+  //! philox_engine.
+  //!
+  //! @param s The seed used to initialize this philox_engine's state.
+  _CCCL_API philox_engine() noexcept
+  {
+    seed(default_seed);
+  }
+  _CCCL_API explicit philox_engine(const result_type __seed) noexcept
+  {
+    seed(__seed);
+  }
+
+  template <class Sseq>
+  _CCCL_API explicit philox_engine(Sseq& q) noexcept
+  {
+    seed(q);
+  }
+
+  //! This method initializes this philox_engine's state, and optionally accepts
+  //! a seed value.
+  //!
+  //! @param __s The seed used to initializes this philox_engine's state.
+  _CCCL_API void seed(result_type __s = default_seed) noexcept
```

Comment:
I am wondering, could we make all of those `constexpr`?

I am dying to get a good constexpr enabled RNG

Reply:
With some small tweaks to mulhilo it works perfectly.